libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route2/ReplayRouteSmoother.kt

```diff
@@ -0,0 +1,193 @@
+package com.mapbox.navigation.core.replay.route2
+
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.geojson.LineString
+import com.mapbox.geojson.Point
+import com.mapbox.turf.TurfConstants
+import com.mapbox.turf.TurfMeasurement
+import com.mapbox.turf.TurfMeasurement.EARTH_RADIUS
+import kotlin.math.abs
+import kotlin.math.cos
+import kotlin.math.sin
+import kotlin.math.sqrt
+
+/**
+ * In order to estimate the speed needed to traverse a [DirectionsRoute] we need
+ * a route with locations that are significant for speed.
+ *
+ * This class reduces a list of location coordinates into a list of locations
+ * with direction changes beyond a threshold.
+ */
+internal class ReplayRouteSmoother {
+
+    /**
+     * Take a list of coordinates and return a list of locations where there are significant
+     * changes in bearing. Each location will have a bearing and a distance.
+     */
+    fun smoothRoute(distinctPoints: List<Point>, thresholdMeters: Double): List<ReplayRouteLocation> {
+        val smoothIndices = smoothRouteIndices(distinctPoints, thresholdMeters)
+        val smoothLocations = smoothIndices.map { ReplayRouteLocation(it, distinctPoints[it]) }
+
+        var segmentStart = smoothLocations[0]
+        var segmentEnd = smoothLocations[1]
+        var bearing = TurfMeasurement.bearing(segmentStart.point, segmentEnd.point)
+        var segmentRoute = segmentRoute(distinctPoints, segmentStart.routeIndex!!, segmentEnd.routeIndex!!)
+        var distance = TurfMeasurement.length(LineString.fromLngLats(segmentRoute), TurfConstants.UNIT_METERS)
+        smoothLocations.first().apply {
+            this.bearing = bearing
+            this.distance = distance
+        }
+
+        for (i in 1 until smoothLocations.size - 1) {
+            segmentStart = smoothLocations[i]
+            segmentEnd = smoothLocations[i + 1]
+            segmentRoute = segmentRoute(distinctPoints, segmentStart.routeIndex!!, segmentEnd.routeIndex!!)
+            distance = TurfMeasurement.length(LineString.fromLngLats(segmentRoute), TurfConstants.UNIT_METERS)
+            bearing = TurfMeasurement.bearing(segmentStart.point, segmentEnd.point)
+
+            segmentStart.bearing = bearing
+            segmentStart.distance = distance
+        }
+
+        // Stop at the end of the route
+        smoothLocations.last().apply {
+            this.bearing = bearing
+            this.distance = 0.0
+        }
+
+        return smoothLocations
+    }
+
+    /**
+     * Given a list of location points, identify indices that are not
+     * within [thresholdMeters] of a straight road.
+     */
+    fun smoothRouteIndices(points: List<Point>, thresholdMeters: Double): List<Int> {
```

Comment:
Same as above re: extracting blocks of code into separate well named `private` functions

Reply:
made private